src/range.cr

```diff
@@ -159,10 +169,19 @@ struct Range(B, E)
     yield end_value if !@exclusive && (begin_value.nil? || !(end_value < begin_value))
     current = end_value
 
-    while begin_value.nil? || begin_value < current
-      current = current.pred
-      yield current
-    end
+    # TODO: The macro interpolations are a workaround until #9324 is fixed.
+
+    {% if B == Nil %}
+      while true
+        current = current.pred
+        {{ "yield current".id }}
+      end
+    {% else %}
+      while begin_value.nil? || begin_value < current
```

Comment:
Is the first condition still possible in this branch?

Reply:
Yes, if E is a union that has nil, and at runtime happens to be nil. E won't be Nil but at runtime it will be nil.